shotover-proxy/src/server.rs

```diff
@@ -27,9 +29,114 @@ impl<T: Decoder<Item = Messages, Error = anyhow::Error> + Clone + Send> CodecRea
 pub trait CodecWriteHalf: Encoder<Messages, Error = anyhow::Error> + Clone + Send {}
 impl<T: Encoder<Messages, Error = anyhow::Error> + Clone + Send> CodecWriteHalf for T {}
 
+pub trait CodecErrorFixup {
+    /// An method to fix up an error.
+    ///
+    /// Codecs that do not have additional error handling should return `(None, None, Optional<Err>)`
```

Comment:
I think an `enum` would be preferable to a combination of `Option`.

Reply:
Why does ' down_msg` exist if we aren't using it?

Reply:
1. yes please, it will be easy to change the type in the future if we need more information from the error.
2. While this is how I would approach the problem, I certainly lack the years of experience that you have!
I would like to hear @benbromhead thoughts on if this abstraction is appropriate.